### PR TITLE
Fix and speed-up norm_backwards

### LINKED ISSUE #3264: F.normalize NaN gradient
0.2.0+dc6510f

`F.normalize(p=1)` gives NaN gradients.
Manually dividing by the sum works.

### Code

```py
import torch
from torch.autograd.variable import Variable
import torch.nn.functional as F

a = Variable(torch.FloatTensor([[0,0,0,0,0.1996]]),requires_grad=True)
# b = F.normalize(a, p=1, dim=1)  # nan nan nan nan 0
# b = a / a.sum().clamp(min=1e-12)  # 0 0 0 0 0
b.sum().backward()

print(a.grad.data)
```
### Output
```
nan nan nan nan 0
```
or
```
0 0 0 0 0
```

### PR DESCRIPTION
**Pull request by colesbury**
Fixes #3264

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -1790,8 +1791,14 @@ class dont_convert(tuple):
...
-    ('norm', (S, S, S), (2,)),
-    ('norm', (S, S, S), (3,), '3'),
+    ('norm', (S, S), (2,)),
+    ('norm', (S, S), (0,), '0'),
+    ('norm', (S, S), (0.5,), '0_5'),
+    ('norm', (S, S), (1,), '1'),
+    ('norm', (S, S), (3,), '3'),
+    ('norm', (S, S), (-1,), 'neg_1'),
+    ('norm', (S, S), (-0.5,), 'neg_0_5'),
+    ('norm', (S, S), (-1.5,), 'neg_1_5'),
```
Comment (apaszke) [inline test/test_autograd.py]: What? Does `p`-norm on values less or equal 0 even make sense? IIRC it's not even a norm when `p < 1` (no triangle inequality) and it seems that when it's negative it actually puts similar vectors far, and different ones close.

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -348,10 +348,10 @@
   self: zeros_like(grad)
 
 - name: norm(Tensor self, Scalar p=2)
-  self: norm_backward(grad, self, p)
+  self: norm_backward(grad, self, p, result)
 
 - name: norm(Tensor self, Scalar p, int64_t dim, bool keepdim=False)
-  self: norm_backward(grad, self, p, dim, keepdim)
+  self: norm_backward(grad, self, p, destination, dim, keepdim)
```
Comment (apaszke) [inline tools/autograd/derivatives.yaml]: Is `result` saved as a `SavedVariable` here, or is it just a scalar?

**Review by apaszke — COMMENTED** [review]
Haven't verified the math, but gradchecks are passing so should be ok

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (colesbury) [reply tools/autograd/derivatives.yaml]: It's saved as a `SavedVariable`

File: test/test_autograd.py (hunk 1 above)
Comment (colesbury) [reply test/test_autograd.py]: As a general rule, if we support something on `Tensor`, we need to make sure the backwards operation computes the derivative correctly.

Comment (colesbury) [reply test/test_autograd.py]: It's still minimized when the vector is small. For a two element vector with p=-2:

image 

[formula](https://www.wolframalpha.com/input/?i=(%7Cx%7C%5E-2%2B%7Cy%7C%5E-2)%5E-0.5)
